impl/src/bitfield_specifier.rs

```diff
@@ -45,17 +45,69 @@ fn generate_or_error(input: TokenStream2) -> syn::Result<TokenStream2> {
     }
 }
 
-fn generate_enum(input: syn::ItemEnum) -> syn::Result<TokenStream2> {
+struct Attributes {
+    bits: Option<usize>,
+}
+
+fn parse_attrs(attrs: &[syn::Attribute]) -> syn::Result<Attributes> {
+    let attributes = attrs
+        .iter()
+        .filter(|attr| attr.path.is_ident("bits"))
+        .fold(
+            Ok(Attributes { bits: None }),
+            |acc: syn::Result<Attributes>, attr| {
+                let mut acc = acc?;
+                if acc.bits.is_some() {
+                    return Err(syn::Error::new(
+                        Span2::call_site(),
+                        "More than one 'bits' attributes is not permitted",
+                    ))
+                }
+                let meta = attr.parse_meta()?;
+                acc.bits = match meta {
+                    syn::Meta::NameValue(syn::MetaNameValue {
+                        lit: syn::Lit::Int(lit),
+                        ..
+                    }) => Some(lit.base10_parse::<usize>()?),
+                    _ => {
+                        return Err(syn::Error::new(
+                            Span2::call_site(),
+                            "could not parse 'bits' attribute",
+                        ))
+                    }
+                };
+                Ok(acc)
+            },
+        )?;
+    Ok(attributes)
+}
+
+pub fn generate_enum(input: syn::ItemEnum) -> syn::Result<TokenStream2> {
+    let attributes = parse_attrs(&input.attrs)?;
     let enum_ident = &input.ident;
-    let count_variants = input.variants.iter().count();
-    if !count_variants.is_power_of_two() {
+
+    let (bits, returns_result) = match attributes.bits {
+        Some(bits) => (bits, true),
+        None => {
+            let count_variants = input.variants.iter().count();
+            if !count_variants.is_power_of_two() {
+                return Err(syn::Error::new(
+                    Span2::call_site(),
+                    "A BitfieldSpecifier with a non-power-of-2 number of variants must have the 'bits' attribute specified",
+                ));
+            }
+            // We can take `trailing_zeros` returns type as the required amount of bits.
+            (count_variants.trailing_zeros() as usize, false)
+        }
+    };
+
+    // We only support up to 32-bit enums because of the EnumError type
+    if bits > 32 {
         return Err(syn::Error::new(
             Span2::call_site(),
-            "BitfieldSpecifier expected a number of variants which is a power of 2",
+            "An enum may not be larger than 32 bits",
         ))
```

Comment:
Can you make use of `format_err_spanned` here please? I see it was still missing from before.

Reply:
Also when you are on it please use the input `enum`'s span here instead of `call_site`. :)